crates/sui-types/src/message_envelope.rs

```diff
@@ -41,6 +42,27 @@ impl<T: Message, S> Envelope<T, S> {
     pub fn into_data(self) -> T {
         self.data
     }
+
+    pub fn into_data_and_sig(self) -> (T, S) {
+        let Self {
+            data,
+            auth_signature,
+            ..
+        } = self;
+        (data, auth_signature)
+    }
+
+    pub fn auth_sig(&self) -> &S {
+        &self.auth_signature
+    }
+
+    pub fn digest(&self) -> &T::DigestType {
+        self.digest.get_or_init(|| self.data.digest())
+    }
+
+    pub fn data_mut_for_testing(&mut self) -> &mut T {
+        &mut self.data
+    }
```

Comment:
from what i understand, the data:T here is not what the signature is committing to, correct? SenderSignedData is used for this generics. it might get confusing what data we are signing on, and perhaps we can replace it with `data: IntentMessage<T>` in the future
```
pub type Transaction = Envelope<SenderSignedData, EmptySignInfo>;
pub type VerifiedTransaction = VerifiedEnvelope<SenderSignedData, EmptySignInfo>;
pub type CertifiedTransaction = Envelope<SenderSignedData, AuthorityStrongQuorumSignInfo>;
```

Reply:
data: T is what the signature is committing to. For instance, in the case of `CertifiedTransaction`, validators are signing the `SenderSignedData`.

Reply:
i see, so if we add epoch ID it should goes into SenderSignedData? perhaps we should rename it (in a separate PR when we add epoch ID)

Reply:
epoch ID should already be in `AuthoritySignInfo` and `AuthorityQuorumSignInfo` in crypto.rs. When validator is signing it will needs to commit to that epoch.
We haven't seen a need for users to commit to an epoch yet (i.e. whether a tx must be executed in a specific epoch).